src/gax/src/error/core_error.rs

```diff
@@ -12,161 +12,659 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use super::CredentialsError;
+use super::rpc::Status;
+use http::HeaderMap;
+
 type BoxError = Box<dyn std::error::Error + Send + Sync>;
 
 /// The core error returned by all client libraries.
+///
+/// The client libraries report errors from multiple sources. For example, the
+/// service may return an error, the transport may be unable to create the
+/// necessary connection to make a request, the request may timeout before a
+/// response is received, the retry policy may be exhausted, or the library may
+/// be unable to format the request due to invalid or missing application
+/// application inputs.
+///
+/// Most applications will just return the error or log it, without any further
+/// action. However, some applications may need to interrogate the error
+/// details. This type offers a series of predicates to determine the error
+/// kind. The type also offers accessors to query the most common error details.
+/// Applications can query the error [source][std::error::Error::source] for
+/// deeper information.
+///
+/// # Example
+/// ```
+/// use google_cloud_gax::error::Error;
+/// match example_function() {
+///     Err(e) if matches!(e.status(), Some(_)) => {
+///         println!("service error {e}, should have a status={:?}", e.status());
```

Comment:
Can you say `if let Some(s) = e.status()`

And then just print `"service error: {s}"`

Reply:
Nope:

```
error[E0658]: `if let` guards are experimental
 --> src/gax/src/error/core_error.rs:42:12
  |
7 |     Err(e) if let Some(s) = e.status() => {
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #51114 <https://github.com/rust-lang/rust/issues/51114> for more information
  = help: you can write `if matches!(<expr>, <pattern>)` instead of `if let <pattern> = <expr>`
```

Reply:
Ack, thanks for checking